[Sema] Always look through optionals for unresolved member lookup

Jumhyn opened the pull request:
Previously, we only looked through optionals on `UnresolvedMemberExpr`s when there were no results from `Optional` itself. However, in cases where `Optional` has a member with the same name, this would prevent programs from compiling even if the member on `Optional` was obviously invalid in the context of the `UnresolvedMemberExpr`. Consider this example:

```
import SwiftUI

extension UIColor {
  static let background: UIColor? = UIColor.black
}

let backgroundColor: UIColor? = .background
```

In SwiftUI, `Optional` conditionally conforms to `View` whenever `Wrapped` conforms to `View`. Additionally, `View` has an extension which provides an instance method named `background(_:alignment:)`. This means that unresolved member lookup of `background` will find the method on `Optional`, and won't continue to look in `UIColor` for the user-defined `background` member.

This PR allows the above code to compile by unconditionally looking into `Wrapped` for unresolved member lookup, even if we already found results in `Optional<Wrapped>`. Because we already prefer overloads found on `Optional` to those found on `Wrapped` in _CSRanking_, situations where there are valid members on both should continue to choose the member on `Optional` (though this definitely needs a source-compat run).

Resolves SR-13815.

Jumhyn committed:
[Sema] Always look through optionals for unresolved member lookup

Jumhyn commented:
cc @xedin @hborla interested in both of your thoughts on this! It seems like this should be source compatible given our ranking logic but I'd like more eyes to make sure I'm not missing something...

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -6682,13 +6682,12 @@ performMemberLookup(ConstraintKind constraintKind, DeclNameRef memberName,
       // If we don't have a wrapped type yet, we can't look through the optional
       // type.
-      if (objectType->getAs<TypeVariableType>()) {
+      if (objectType->getAs<TypeVariableType>() && result.ViableCandidates.empty()) {
```
I think this would break source compatibility when score of something found on `Wrapped` be better than score on `Optional` itself...

Jumhyn replied:
Since we already score `Decl` better than `DeclViaUnwrappedOptional`, the score on `Wrapped` would have to outweigh the disadvantage from the `OverloadChoiceKind` mismatch. Do you think such situations are likely to occur in practice?

Is there any way to unconditionally prefer the non-optional choices, in the case where we've found multiple solutions?

xedin replied:
> Since we already score Decl better than DeclViaUnwrappedOptional

Solutions are first ranked on the score produced by the solver (https://github.com/apple/swift/blob/main/lib/Sema/CSRanking.cpp#L758) and then their declaration diff is compared if scores match. It's quite possible that choices could have different optionality of require other implicit conversion which would change the behavior here.

Jumhyn replied:
@xedin Ah, thank you for pointing that out. The situation where we have identically-named static members on both `Optional<Wrapped>` and `Wrapped` where both properly type-check seems pretty unlikely to me, but if it's above the risk level for being considered source-compatible (as opposed to just a bug) then so be it.

Is there a way to add the `Wrapped` overloads but not skip over them whenever any `Optional` overloads successfully type check?

xedin replied:
One way to implement this without changing existing behavior would be to lift `DeclViaUnwrappedOptional` into the solver score to make sure regular declarations are always preferred over the ones found on wrapped type.

Jumhyn commented:
@xedin I've taken your suggestion and lifted `DeclViaUnwrappedOptional` into `ScoreKind` for `UnresolvedMemberExpr` overload resolution. I've also added a couple tests to make sure that we resolve to overloads on `Optional` when there are matching overloads on `Wrapped` that would otherwise be preferred. Let me know what you think!

xedin reviewed (approved):
LGTM! I think might be worth it to add a unit test for this behavior as it might be much easier to make sure the behavior is correct without having to rely on debug output formatting of the solver.

LucianoPAlmeida commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -2794,6 +2794,11 @@ void ConstraintSystem::resolveOverload(ConstraintLocator *locator,
     increaseScore(SK_DisfavoredOverload);
   }
+
+  if (choice.getKind() == OverloadChoiceKind::DeclViaUnwrappedOptional &&
```
@Jumhyn Does this affects situations like https://bugs.swift.org/browse/SR-11535 in some way?

xedin replied:
It would be good to add a test-case from that SR to the suite.

Jumhyn replied:
I don't think so, in that situation the `Optional.none` overload doesn't provide a valid solution since the generic argument to `Foo` cannot be inferred, and then we find the `Foo.none` solution when we attempt to bind the result of the `UnresolvedMemberExpr` to `Foo` (rather than `Foo?`). I'll verify that the behavior is the same, though.

LucianoPAlmeida replied:
Yeah, seems like you are correct! I didn't note that. So it means this SR isn't really a bug...

LucianoPAlmeida replied:
But maybe we can add this specific case to the suit and resolve it? WDYT? I can do this in a separated PR no problem :)

Jumhyn replied:
I’m happy to add it as part of this PR, but probably won’t have a chance to make edits until tomorrow evening at the earliest. If you want to put up a PR before then, feel free!

LucianoPAlmeida replied:
It's there #34744 
Actually, I found that we already have the test case added, so I just adjusted and added an explanation comment :)

Jumhyn commented:
@xedin Yeah that makes sense! Is there a good example I could refer to about how to do that? I haven't written a test like that before.

xedin commented:
I added a couple of unit tests which use `ConstraintSystem` and related types in `unittests/Sema`. You’d need to create a couple of fake declarations, generate constraints for a fake unresolved member reference and check that solution has an expected declaration picked and score doesn’t have member-via-unwrap increased.

Jumhyn commented:
Alright, I'll give those a look—thanks! Will ping you when this PR is ready for another look.

xedin commented:
Thank you!

Jumhyn commented:
@xedin Finally got a chance to sit down with this and realized that relying on the solver output is overkill. All we really need is the type-checked AST, so I've switched to just checking the output of `-dump-ast` instead. Do you think that's sufficient, or do you still want a unit test here as well?

xedin commented:
@Jumhyn I'd still prefer if we had that as a unit test but I think it's okay to use `-dump-ast` to verify that, at least it's more stable than `-debug-constraints`.

LucianoPAlmeida commented:
@Jumhyn @xedin After those changes member lookup here https://github.com/apple/swift/blob/a9f5107b44c71929521ffbf3a91c5081554f272d/lib/Sema/CSSimplify.cpp#L7183 will return both candidates for Optional and Wrapped types? Just checking as I'm trying to change the CSApply warning to a fix and right now we have to do another lookup ...

Jumhyn commented:
@LucianoPAlmeida Yes, the results from that line should include candidates from both `Optional` and `Wrapped` (but only when we’re doing lookup for an `UnresolvedMemberExpr`).

LucianoPAlmeida commented:
> @LucianoPAlmeida Yes, the results from that line should include candidates from both `Optional` and `Wrapped` (but only when we’re doing lookup for an `UnresolvedMemberExpr`).

Great thanks, I think that could simplify things in that case :)

Jumhyn commented:
@LucianoPAlmeida Unfortunately, I probably won't have a chance for the next couple of days to write the unit test to get this mergeable, so you may have to wait for a bit for the changes here to be usable for your fix.

LucianoPAlmeida commented:
> @LucianoPAlmeida Unfortunately, I probably won't have a chance for the next couple of days to write the unit test to get this mergeable, so you may have to wait for a bit for the changes here to be usable for your fix.

No worries @Jumhyn :) 
Those are very simple changes so I've already "cherry-picked" to my branch to work with them. It did break the implementation that I had already so it will take some time for me to have something also.

LucianoPAlmeida commented:
@xedin @Jumhyn Quick question, in this case 
```
enum Foo {
  case bar
  case none
}

let _: Foo?? = .none
```
Name lookup is actually returning only 2 viable candidates where the base is `Foo??` and `Foo?`, so my question is if that is expected and intentional or it should have dig deeper and find also base `Foo`?

xedin commented:
@LucianoPAlmeida It intentionally only unwrap a single layer of optionality at the moment.

LucianoPAlmeida commented:
> @LucianoPAlmeida It intentionally only unwrap a single layer of optionality at the moment.

Ah, thank you @xedin :) 
I think that will require another lookup then in order to record the fix correctly then

xedin commented:
That would be okay since it's performed in diagnostic mode anyway.

LucianoPAlmeida commented:
> That would be okay since it's performed in diagnostic mode anyway.

Humm, that may be a problem because since it is a warning we have to always record it ... but on the other hand the extra lookup will be done only in certain situations involving deep optionals. I'm finishing up the changes and pushing, so we can discuss it there :)

Jumhyn committed:
[Unit tests] Add unit tests for optional lookup behavior

Jumhyn commented:
Thank you for your patience @xedin! Finally had a chance to sit down an write up the unit test(s). Let me know if this matches what you were picturing.

cc @LucianoPAlmeida, this will hopefully be merge-ready soon! :)

xedin commented on unittests/Sema/SemaFixture.cpp:
```diff
@@ -76,6 +76,39 @@ Type SemaTest::getStdlibType(StringRef name) const {
+  if (results.size() != 1)
+    return nullptr;
+
+  if (auto *NTD = dyn_cast<NominalTypeDecl>(results.front()))
```
This could be simplified down to `return dyn_cast...`

xedin reviewed (approved):
Looks great! That’s exactly what I had in mind.

Jumhyn commented:
@xedin Updated per your comment! Would you mind kicking off the tests on this?

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

LucianoPAlmeida commented on UnresolvedMemberLookupTests.cpp:
```diff
@@ -0,0 +1,82 @@
+//===--- ConstraintSimplificationTests.cpp --------------------------------===//
```
Is this the correct name? 
Also, the GitHub UI shows no path so is this file on the root of the repo? Should this be in `/unittests/Sema/` folder?

Jumhyn replied:
Womp womp, bad copy-paste job. Will fix this up!

LucianoPAlmeida commented on unittests/Sema/SemaFixture.cpp:
```diff
@@ -76,6 +76,36 @@ Type SemaTest::getStdlibType(StringRef name) const {
   return Type();
 }
 
+NominalTypeDecl *SemaTest::getStdlibNomicalTypeDecl(StringRef name) const {
```
Typo: `Nomical` -> `Nominal`? :)

LucianoPAlmeida commented on unittests/Sema/SemaFixture.cpp:
```diff
@@ -76,6 +76,36 @@ Type SemaTest::getStdlibType(StringRef name) const {
   return Type();
 }
 
+NominalTypeDecl *SemaTest::getStdlibNomicalTypeDecl(StringRef name) const {
```
Nit: If this method was `getNominalTypeDecl(Module&, StringRef)` that could be more general for use in the future if there were to be multimodule unit tests. We could even `getStdlibNominalTypeDecl` that gets the stdlib module and call the generic one ... but it is just an idea not really sure it would be useful 
@xedin does this make sense?

xedin replied:
I think that could be generalized later once we have found a use-case for it...

LucianoPAlmeida replied:
Make sense, thanks :)

LucianoPAlmeida commented on UnresolvedMemberLookupTests.cpp:
```diff
@@ -0,0 +1,82 @@
+  auto *intMember = addExtensionVarMember(intTypeDecl, "test", intOptType);
+  addExtensionVarMember(optTypeDecl, "test", stringType);
+
+  auto *UME = new (Context) UnresolvedMemberExpr(SourceLoc(), DeclNameLoc(), DeclNameRef(Context.getIdentifier("test")),
```
Nit: Should this be clang-formatted? Seems like it's not in the 80 col rule :)

Jumhyn replied:
You are correct :)

Jumhyn commented:
@LucianoPAlmeida Latest push addresses your comments and fixes the test failure on Windows (in theory🤞). From what I could tell the Linux and macOS test failures were not related to the code touched here... mind kicking off another round of tests?

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

LucianoPAlmeida commented:
@swift-ci please test MacOS Platform

LucianoPAlmeida commented:
@swift-ci Please test OS X platform

Jumhyn commented:
Windows failure looks like a flake. Rebased against `main` to see if that resolves the (seemingly unrelated) macOS failure... @xedin @LucianoPAlmeida one more round of tests, please?

LucianoPAlmeida commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

Jumhyn commented:
@xedin All checks passed!
